### ISSUE TITLE
[YCQL] Need to initialize permission_ in the PTGrantRevokePermission constructor

### ISSUE THREAD

**Issue by mbautin** [yugabyte/yugabyte-db#5577]
Jira Link: [DB-2087](https://yugabyte.atlassian.net/browse/DB-2087)
PVS Studio found that we are not setting the permission_ field here in this constructor:
![image](https://user-images.githubusercontent.com/552936/92045427-5b5b9680-ed35-11ea-972d-38e34f9234b4.png)

```c++
PTGrantRevokePermission::PTGrantRevokePermission(MemoryContext* memctx,
                                                 YBLocation::SharedPtr loc,
                                                 GrantRevokeStatementType statement_type,
                                                 const MCSharedPtr<MCString>& permission_name,
                                                 const ResourceType& resource_type,
                                                 const PTQualifiedName::SharedPtr& resource_name,
                                                 const PTQualifiedName::SharedPtr& role_name)
    : TreeNode(memctx, loc),
      statement_type_(statement_type),
      permission_name_(permission_name),
      complete_resource_name_(resource_name),
      role_name_(role_name),
      resource_type_(resource_type) {
}
```

(in `pt_grant_revoke.cc`).

Not sure what the right default value would be here.

**Comment by hectorgcr** [comment]
We currently set `permission_` in the `Analyze` method:

```
auto iterator = kPermissionMap.find(string(permission_name_->c_str()));
  if (iterator == kPermissionMap.end()) {
    return sem_context->Error(this, Substitute("Unknown Permission '$0'",
                                               permission_name_->c_str()).c_str(),
                              ErrorCode::SYNTAX_ERROR);
  }

  permission_ = iterator->second;
```

I guess we could create a new PermissionType to indicate an invalid or unset permission if we really want to set `permission_` in the constructor.

**Comment by tedyu** [comment]
I think part of the reason permission_ is assigned in Analyze() is that ctor doesn't have return value (since the lookup of permission_name_ in the map may find nothing).

We can either:

1. add a comment in ctor explaining this choice
2. search the map in ctor. If the permission_name_ is found, assign permission_. Otherwise assign a new enum (e.g. PermissionType::UNKNOWN) as Hector suggested. In the Analyze(), we just check whether permission_ is unknown.

BTW, it seems we can replaced the map with unordered_map.
See https://www.geeksforgeeks.org/map-vs-unordered_map-c/

**Comment by tedyu** [comment]
As far as I can tell, permission_ field is not accessed prior to the search of permission_name_.

Considering the following definition:

enum PermissionType {
  ALTER_PERMISSION = 0,

I would think that implicitly permission_ has default value of ALTER_PERMISSION.
Assigning permission_ (an explicit) UNKNOWN value would be better than leaving it with default of ALTER_PERMISSION.

**Comment by tedyu** [comment]
Actually since permission_ field is not accessed outside of Analyze(), we can make it local to Analyze() - after all, it doesn't convey more information than permission_name_.

**Comment by tedyu** [comment]
The simplest option is to initialize permission_ in the ctor with PermissionType::ALL_PERMISSION (as is done in header file now).

**Comment by frozenspider** [comment]
Fixed in 0a50f0b7bd73b083317587a6a80f9253e19337bf
